File: src/main/java/de/jarox/gommemode/entity/GommeEntity.java
package de.jarox.gommemode.entity;

import net.minecraft.client.MinecraftClient;
import net.minecraft.command.argument.EntityAnchorArgumentType;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.LivingEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.Arm;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;

public class GommeEntity extends LivingEntity {

    public GommeEntity(EntityType<? extends LivingEntity> entityType, World world) {
        super(entityType, world);
    }

    @Override
    public Iterable<ItemStack> getArmorItems() {
        return null;
    }

    @Override
    public ItemStack getEquippedStack(EquipmentSlot slot) {
        return new ItemStack(Items.DIAMOND_SWORD);
    }

    @Override
    public void equipStack(EquipmentSlot slot, ItemStack stack) {
    }

    @Override
    public Arm getMainArm() {
        return Arm.RIGHT;
    }

    @Override
    public void baseTick() {
        super.baseTick();
        lookAtPlayer();
    }

    private void lookAtPlayer() {
        if (this.getWorld().isClient) {
            assert MinecraftClient.getInstance().player != null;
            Vec3d playerPos = MinecraftClient.getInstance().player.getPos();
            this.lookAt(EntityAnchorArgumentType.EntityAnchor.FEET, playerPos);
        }
    }
}

File: src/main/kotlin/de/jarox/gommemode/entity/GommeEntity.kt
package de.jarox.gommemode.entity

import net.minecraft.client.Minecraft
import net.minecraft.commands.arguments.EntityAnchorArgument
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.HumanoidArm
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.level.Level

/**
 * A custom living entity that always holds a diamond sword and faces the local player.
 *
 * @param entityType The entity type for this entity
 * @param level The world this entity exists in
 */
class GommeEntity(
    entityType: EntityType<out LivingEntity>,
    level: Level,
) : LivingEntity(entityType, level) {
    private val diamondSword: ItemStack = ItemStack(Items.DIAMOND_SWORD)

    /**
         * Provide the entity's equipped item for the given equipment slot.
         *
         * Only the main hand returns the entity's diamond sword; all other slots return `ItemStack.EMPTY`.
         *
         * @param slot The equipment slot to query.
         * @return `diamondSword` when `slot` is `EquipmentSlot.MAINHAND`, `ItemStack.EMPTY` otherwise.
         */
        override fun getItemBySlot(slot: EquipmentSlot): ItemStack =
        when (slot) {
            EquipmentSlot.MAINHAND -> diamondSword
            else -> ItemStack.EMPTY
        }

    override fun setItemSlot(
        slot: EquipmentSlot,
        stack: ItemStack,
    ) {
        // Intentionally no-op: this entity's equipment is managed by getItemBySlot
    }

    override fun getMainArm(): HumanoidArm = HumanoidArm.RIGHT

    override fun baseTick() {
        super.baseTick()
        faceLocalPlayer()
    }

    /**
     * Rotates this entity to face the local player's feet position.
     * Only executes on the client side where the player reference is available.
     */
    private fun faceLocalPlayer() {
        if (!level().isClientSide) return

        val player = Minecraft.getInstance().player ?: return
        lookAt(EntityAnchorArgument.Anchor.FEET, player.position())
    }
}
